storage/provider/openstack.go

```diff
@@ -0,0 +1,328 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package provider
+
+import (
+	"math"
+	"net/url"
+
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/storage"
+
+	"github.com/juju/errors"
+	"github.com/juju/names"
+
+	"gopkg.in/goose.v1/cinder"
+	"gopkg.in/goose.v1/client"
+	"gopkg.in/goose.v1/identity"
+	"gopkg.in/goose.v1/nova"
+)
+
+const (
+	OpenstackProviderType = storage.ProviderType("openstack")
+)
+
+type openstackProvider struct {
+	clientFactory func(*storage.Config) (OpenstackAdapter, error)
+}
+
+var _ storage.Provider = (*openstackProvider)(nil)
+
+// VolumeSource implements storage.Provider.
+func (p *openstackProvider) VolumeSource(environConfig *config.Config, providerConfig *storage.Config) (storage.VolumeSource, error) {
+
+	if err := p.ValidateConfig(providerConfig); err != nil {
+		return nil, err
+	}
+
+	openstackClient, err := p.clientFactory(providerConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &openstackVolumeSource{openstackClient}, nil
+}
+
+// FilesystemSource implements storage.Provider.
+func (p *openstackProvider) FilesystemSource(environConfig *config.Config, providerConfig *storage.Config) (storage.FilesystemSource, error) {
+	return nil, errors.NotSupportedf("filesystems")
+}
+
+// Supports implements storage.Provider.
+func (p *openstackProvider) Supports(kind storage.StorageKind) bool {
+	switch kind {
+	case storage.StorageKindBlock:
+		return true
+	}
+	return false
+}
+
+// Scope implements storage.Provider.
+func (s *openstackProvider) Scope() storage.Scope {
+	panic("not implemented")
```

Comment:
Are you sure? I just updated from trunk and the interface specifies Scope. Should I be rebasing from somewhere else?